src/apply_gridshift.cpp

```diff
@@ -334,33 +289,92 @@ PJ_LP proj_hgrid_value(PJ *P, PJ_LP lp) {
     return out;
 }
 
-PJ_LP proj_hgrid_apply(PJ *P, PJ_LP lp, PJ_DIRECTION direction) {
-    struct CTABLE *ct;
-    int inverse;
+static
+PJ_LP proj_hgrid_apply_internal(PJ_CONTEXT* ctx,
+                       PJ_LP lp,
+                       PJ_DIRECTION direction,
+                       const ListOfHGrids& grids)
+{
     PJ_LP out;
 
-    out.lam = HUGE_VAL; out.phi = HUGE_VAL;
+    out.lam = HUGE_VAL;
+    out.phi = HUGE_VAL;
 
-    ct = find_ctable(P->ctx, lp, P->gridlist_count, P->gridlist);
-
-    if (ct == nullptr || ct->cvs == nullptr) {
-        if( P->gridlist_count == 1 &&
-            strcmp(P->gridlist[0]->gridname, "null") == 0) {
-          // TODO: remove this particular case that is put there just to be
-          // able to handle longitudes outside of -180,180
-          out = lp;
-        } else {
-          pj_ctx_set_errno( P->ctx, PJD_ERR_FAILED_TO_LOAD_GRID );
-        }
+    const auto grid = findGrid(grids, lp);
+    if( !grid ) {
+        pj_ctx_set_errno( ctx, PJD_ERR_FAILED_TO_LOAD_GRID );
         return out;
     }
+    if( grid->isNullGrid() )
+    {
+        return lp;
+    }
 
-    inverse = direction == PJ_FWD ? 0 : 1;
-    out = nad_cvt(lp, inverse, ct);
+    int inverse = direction == PJ_FWD ? 0 : 1;
+    out = nad_cvt(lp, inverse, grid);
 
     if (out.lam == HUGE_VAL || out.phi == HUGE_VAL)
-        pj_ctx_set_errno(P->ctx, PJD_ERR_GRID_AREA);
+        pj_ctx_set_errno(ctx, PJD_ERR_GRID_AREA);
 
     return out;
+}
+
+PJ_LP proj_hgrid_apply(PJ *P, const ListOfHGrids& grids, PJ_LP lp, PJ_DIRECTION direction) {
```

Comment:
`proj_hgrid_apply` is only used internally, so you are allowed to change its function parameters etc. Seems better than to add the `_internal` function (unless there's something I haven't understood).

The `proj_` prefix is misleading. Should be changed to `pj_` here as well.